pkg/controller/direct/eventarc/channel_controller.go

```diff
@@ -0,0 +1,294 @@
+		return mapCtx.Err()
+	}
+
+	paths := []string{}
+	if resource.Provider != a.actual.Provider {
+		paths = append(paths, "provider")
+	}
+	if resource.CryptoKeyName != a.actual.CryptoKeyName {
+		paths = append(paths, "crypto_key_name")
+	}
+
+	var updated *pb.Channel
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
+		// even though there is no update, we still want to update KRM status
+		updated = a.actual
+	} else {
+		resource.Name = a.id.String() // we need to set the name so that GCP API can identify the resource
+		req := &pb.UpdateChannelRequest{
+			Channel:    resource,
+			UpdateMask: &fieldmaskpb.FieldMask{Paths: paths},
+		}
+
+		op, err := a.gcpClient.UpdateChannel(ctx, req)
+		if err != nil {
+			return fmt.Errorf("updating eventarc channel %s: %w", a.id.String(), err)
+		}
+
+		updated, err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting for eventarc channel update %s: %w", a.id.String(), err)
+		}
+		log.V(2).Info("successfully updated eventarc channel", "name", a.id)
+	}
+
+	status := &krm.EventarcChannelStatus{}
+	status.ObservedState = EventarcChannelObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(updated.Name)
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+// Export maps the GCP object to a Config Connector resource `spec`.
+func (a *channelAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.EventarcChannel{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(EventarcChannelSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	obj.Spec.ProjectRef = &refs.ProjectRef{External: a.id.Parent().ProjectID}
+	obj.Spec.Location = a.id.Parent().Location
+	obj.Spec.ResourceID = direct.LazyPtr(a.id.ID())
+
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	u.SetName(a.id.String())
+	u.SetNamespace(obj.Namespace) // This is required KCC controller code convention
+	u.SetGroupVersionKind(krm.EventarcChannelGVK)
+	u.Object = uObj
+
+	return u, nil
+}
+
+// Delete the resource from GCP service when the corresponding Config Connector resource is deleted.
+func (a *channelAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("deleting eventarc channel", "name", a.id)
+
+	req := &pb.DeleteChannelRequest{Name: a.id.String()}
+	op, err := a.gcpClient.DeleteChannel(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			log.V(2).Info("eventarc channel not found", "name", a.id)
+			return false, nil // Resource is gone, consider the delete successful.
+		}
+		if apierrors.IsNotFound(err) { // Handle potential K8S not found errors if applicable
```

Comment:
Could you help me understand why would this happen? I think we don't touch the K8s resource in this function.

Reply:
that wasnt necessary. removed.